Lombiq.JsonEditor/Controllers/AdminController.cs

```diff
@@ -0,0 +1,105 @@
+using OrchardCore.Title.ViewModels;
+using System.Threading.Tasks;
+using YesSql;
+
+namespace Lombiq.JsonEditor.Controllers;
+
+public class AdminController : Controller
+{
+    private readonly IAuthorizationService _authorizationService;
+    private readonly IContentManager _contentManager;
+    private readonly ILayoutAccessor _layoutAccessor;
+    private readonly ISession _session;
+    private readonly IShapeFactory _shapeFactory;
+    private readonly IStringLocalizer<AdminController> T;
+
+    public AdminController(
+        IAuthorizationService authorizationService,
+        IContentManager contentManager,
+        ILayoutAccessor layoutAccessor,
+        ISession session,
+        IShapeFactory shapeFactory,
+        IStringLocalizer<AdminController> stringLocalizer)
+    {
+        _authorizationService = authorizationService;
+        _contentManager = contentManager;
+        _layoutAccessor = layoutAccessor;
+        _session = session;
+        _shapeFactory = shapeFactory;
+        T = stringLocalizer;
+    }
+
+    public async Task<IActionResult> Edit(string contentItemId)
+    {
+        if (string.IsNullOrWhiteSpace(contentItemId) ||
+            await _contentManager.GetAsync(contentItemId, VersionOptions.Latest) is not { } contentItem ||
+            !await CanEditAsync(contentItem))
+        {
+            return NotFound();
+        }
+
+        var titleShape = await _shapeFactory.CreateAsync<TitlePartViewModel>("TitlePart", model =>
+        {
+            model.Title = T["Edit {0} as JSON", contentItem.ContentType];
+            model.ContentItem = contentItem;
+        });
+        await _layoutAccessor.AddShapeToZoneAsync("Title", titleShape);
+
+        return View(new EditContentItemViewModel(contentItem, JsonConvert.SerializeObject(contentItem)));
+    }
+
+    [ValidateAntiForgeryToken]
+    [HttpPost, ActionName(nameof(Edit))]
+    public async Task<IActionResult> EditPost(
+        string contentItemId,
+        string json,
+        string returnUrl,
+        [Bind(Prefix = "submit.Publish")] string submitPublish)
+    {
+        if (string.IsNullOrWhiteSpace(contentItemId) ||
+            string.IsNullOrWhiteSpace(json) ||
+            JsonConvert.DeserializeObject<ContentItem>(json) is not { } contentItem)
+        {
+            return NotFound();
+        }
+
+        if (string.IsNullOrWhiteSpace(contentItem.ContentItemId)) contentItem.ContentItemId = contentItemId;
+        if (!await CanEditAsync(contentItem))
+        {
+            return NotFound();
+        }
+
+        if (await _contentManager.GetAsync(contentItem.ContentItemId, VersionOptions.Latest) is { } existing)
+        {
+            existing.Latest = false;
+            existing.Published = false;
+            _session.Save(existing);
+            contentItem.ContentItemVersionId = null;
+        }
+
+        await _contentManager.PublishAsync(contentItem);
+        _session.Save(contentItem);
```

Comment:
Something similar should happen with this too:

```c#
existing = await _contentManager.GetAsync(contentItem.ContentItemId, VersionOptions.DraftRequired);
...
await _contentManager.PublishAsync(contentItem);

Reply:
This is not good. If the content is versionable, this creates a new draft clone of the content item we have no use for. We don't want that, because the new version's content should come from the JSON, not through cloning the existing item.

Reply:
When the content is versionable, we need to edit a new or existing draft, not the latest version, what might be a published one. Unless you want to let people circumvent versioning, what I wouldn't do.

Reply:
We are not editing the latest version. We are unpublishing it and the publishing a new content item that's deserialized from the received JSON.

Reply:
Why is this necessary instead of utilizing Orchard's versioning?

Reply:
It's not clear to me how that would be possible. The JSON we get from the editor is the next version of the document. I'm not aware of any way to deserialize a JSON document onto an existing C# object. (Even if it's possible in Newtonsoft.Json, Jean-Thierry is now actively working in migrating to System.Text.Json, where I know this is explicitly not possible.) So the draft item created by `VersionOptions.DraftRequired` is completely unusable in this scenario.

Reply:
OK then, let's keep it!

Reply:
Just for info and from memory

- With `STJ` we did a `JConvert` helper to serialize /deserialize a content item as before.

      JConvert.DeserializeObject<ContentItem>(json)

- Yes the `Version` column is not related to the `ContentItemVersionId`, it is used when we save an item with `checkConcurrency: true`, it throws if a given record is written concurrently.

- Yes this is not obvious to call the right content manager methods and in the right order ;)

     First I assume that the editor doesn't allow to modify some data that are still managed internally, like `ContentItemId`, `ContentItemVersionId`, `Latest`, `Published` and so on. And I saw that from the editor we only can publish, not save as a draft.

    Why not if it is an advanced feature intended to act directly on a given version record, but yes here we don't follow the version management pattern, e.g. loading / activating a new version if the type is `Versionable`. Yes, the `Latest` may not be `Published`, hmm but if this is the one we are editing it will be, if an item is already `Published`, `PublishAsync()` does nothing, that why I think you needed to call `_session.Save()`.

    Okay there is no driver validations but as I can see no updating and validating handlers will be called too. So what I suggest is to follow the pattern used in the `ApiController.Post` of the `OC.Contents` module.

    It requests to load and activate a `DraftRequired`, a new version is built if the type is `Versionable`, the result being unpublished, if it doesn't exists here you can return `NotFound`, then it merges this existing item version with the provided one, here can be deserialized from the editor, then it calls `UpdateAsync()` and `ValidateAsync()` that calls validating handlers, finally it does `SaveDraftAsync()` or as here `PublishAsync()` (here the item is not already published) that may unpublish a previous version.

    The whole without having to call any `_session.Save()`.

Reply:
Thank you for chiming in, JT. Reopening this convo, then.

Reply:
I tried just passing the content item to `ApiController.Post` and it works great! Now that it's trivial to handle drafts by just passing true to the second parameter, I've added that too.
The only tricky part was that this action requires `Permissions.AccessContentApi` permission, but it's not so hard to grant that to a temporary claims principal, so I did that.

Reply:
Okay I understand the context so feel free to do whatever you think is better, these are only suggestions.

- We already have a dedicated converter allowing to serialize / deserialize a content item, you are using it here and it will be the same with STJ. Hmm maybe in the controller action you could directly bind to a content item as done in the Api controller action.

- Okay I understand, it is more like an advanced feature allowing to update a given version record without having to deal with validations, draft versions and so on. I was not sure but yes in that case my suggestions are less relevant and maybe what you first did was better.

- Good idea to use the Api action but I would not recommend this pattern, I think it is better to keep a dedicated controller action that would be easier to tweak.

- Otherwise, following my first bad idea, once you have a deserialized content item, could be as simple as the following.

      ...
      var contentItem = await _contentManager.GetAsync(
          deserialized.ContentItemId,
          VersionOptions.DraftRequired);
      if (contentItem is null)
      {
          return NotFound();
      }

      if (!await _authorizationService.AuthorizeAsync(
          User,
          CommonPermissions.EditContent,
          contentItem))
      {
          return Forbid();
      }

      contentItem.Merge(deserialized, _updateJsonMergeSettings);
      await _contentManager.UpdateAsync(contentItem);

      var result = await _contentManager.ValidateAsync(contentItem);
      if (!result.Succeeded)
      {
          // Whatever is better to do.
      }

      await _contentManager.PublishAsync(contentItem);
      ...